jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c

```diff
@@ -317,180 +318,129 @@ static ecma_value_t
 ecma_builtin_string_prototype_object_match (ecma_value_t this_to_string_value, /**< this argument */
                                             ecma_value_t regexp_arg) /**< routine's argument */
 {
-  ecma_value_t regexp_value = ECMA_VALUE_EMPTY;
+#if ENABLED (JERRY_ES2015)
+  if (!(ecma_is_value_undefined (regexp_arg) || ecma_is_value_null (regexp_arg)))
+  {
+    ecma_object_t *obj_p = ecma_get_object_from_value (ecma_op_to_object (this_to_string_value));
+    ecma_value_t matcher = ecma_op_object_get_by_symbol_id (obj_p, LIT_MAGIC_STRING_MATCH);
+    ecma_deref_object (obj_p);
 
-  ecma_value_t ret_value = ecma_builtin_string_prepare_search (regexp_arg, &regexp_value);
+    if (ECMA_IS_VALUE_ERROR (matcher))
+    {
+      return matcher;
+    }
 
-  if (ECMA_IS_VALUE_ERROR (ret_value))
-  {
-    return ret_value;
-  }
+    if (!ecma_is_value_undefined (matcher) && !ecma_is_value_null (matcher))
+    {
+      if (!ecma_op_is_callable (matcher))
+      {
+        ecma_free_value (matcher);
+        return ecma_raise_type_error (ECMA_ERR_MSG ("@@match is not callable"));
+      }
 
-  JERRY_ASSERT (!ecma_is_value_empty (regexp_value));
-  ecma_object_t *regexp_obj_p = ecma_get_object_from_value (regexp_value);
+      ecma_object_t *matcher_method = ecma_get_object_from_value (matcher);
 
-  /* 5. */
-  ecma_value_t global_value = ecma_op_object_get_by_magic_id (regexp_obj_p, LIT_MAGIC_STRING_GLOBAL);
+      ecma_value_t arguments[] = { this_to_string_value };
+      ecma_value_t result = ecma_op_function_call (matcher_method, regexp_arg, arguments, 1);
 
-  if (ECMA_IS_VALUE_ERROR (global_value))
-  {
-    ecma_deref_object (regexp_obj_p);
-    return global_value;
+      ecma_deref_object (matcher_method);
+      return result;
+    }
   }
 
-  ecma_string_t *this_string_p = ecma_get_string_from_value (this_to_string_value);
-  ecma_ref_ecma_string (this_string_p);
+  ecma_value_t result = ECMA_VALUE_ERROR;
 
-  JERRY_ASSERT (ecma_is_value_boolean (global_value));
+  ecma_string_t *str_p = ecma_op_to_string (this_to_string_value);
 
-  if (ecma_is_value_false (global_value))
+  if (JERRY_UNLIKELY (str_p == NULL))
   {
-    /* 7. */
-    ret_value = ecma_regexp_exec_helper (regexp_value, this_to_string_value, false);
-    ecma_deref_ecma_string (this_string_p);
-    ecma_deref_object (regexp_obj_p);
-    return ret_value;
+    return ECMA_VALUE_ERROR;
   }
 
-  /* 8.a. */
-  ecma_string_t *index_zero_string_p = ecma_get_ecma_string_from_uint32 (0);
+  ecma_value_t regexp_construct_argument[] = { regexp_arg };
+  ecma_value_t new_regexp = ecma_builtin_regexp_dispatch_construct (regexp_construct_argument, 1);
```

Comment:
The RegExp construction code could be shared between ES5.1 and ES6. Could you change the guards so that code is not duplicated? Preferably by keeping the current ES5.1 code path.